[Myat Htet Kyaw] iP

myathk opened the pull request:
# T-800
> "Please give me sensible commands..." -  You dear Terminator

T-800 helps you manage your tasks. It, 
  * supports tasks with repetition
  * is easy to pick up with thorough error handling
  * is fast to use without the need to click
 
You really should,
  1. let T-800 into your life from [this link](https://github.com/NUSmhk/ip/releases/tag/A-Jar)! 
  2. start using it to manage your tasks fast and efficiently
  3. learn it well enough to not piss T-800 off too much

And it is **FREE**! 😃 

Features: 
  - [x] Managing Tasks
  - [x] Managing Deadlines
  - [x] Adding Tasks with repetition

If you LOVE Java OOP and want to improve T-800, here's the `main`  method:

```javascript
public static void main(String[] args) {
        new Duke("./Duke.txt").run();
    }
```

Commit:
Add Gradle support

myathk committed:
Level 1. Greet, Echo, Exit

myathk committed:
Level 2. Add, List

myathk committed:
Level 3. Mark as Done

myathk committed:
Implemented multiple classes Todo, Deadline and Event that inherit from Task class.

myathk committed:
Level 4. ToDos, Events, Deadlines

myathk committed:
Set up semi-automated testing of Duke

myathk committed:
Added error handling for invalid text input and empty descriptions

myathk committed:
Added error handling for empty deadline of deadlines and empty duration of events. Updated input.txt.

myathk committed:
Modified List class to use  ArrayList<Task> to store tasks instead.

myathk committed:
Add support for deleting tasks from the list.

myathk committed:
Add saving task to text file in hard disk and loading from there

myathk committed:
Fix bug in file path

myathk committed:
Add allowing Duke to detect dates and times and print them out in different format

myathk committed:
Revert "Add allowing Duke to detect dates and times and print them out in different format"

This reverts commit 40e78311a45637358a83cae755342b381002d04d.

myathk committed:
Add allowing Duke to detect dates and times and print them out in different format"

myathk committed:
Merge branch 'branch-Level-7'

myathk committed:
Merge branch 'branch-Level-8'

myathk committed:
Fix bugs to let Duke read from text file and also detect Dates/Times

myathk committed:
Add working TaskList and Storage as refactoring for OOP

myathk committed:
Add working Parser and Ui as refactoring for OOP

myathk committed:
Refactor into different Command classes and change main logic of Duke

myathk committed:
Organize all classes into package duke and Command classes inside into package duke.command

myathk committed:
Fix bug parsing text file where there is index out of bounds error

myathk committed:
Add JUnit tests for classes Deadline, Parser and ToDo

myathk committed:
Package Duke as a JAR file

myathk committed:
Add JavaDoc comments for all non-private classes/methods not under duke.command package

myathk committed:
Tweak code to comply with SE coding standard given

myathk committed:
Add Duke function to find task through a search keyword

myathk committed:
Merge branch 'branch-A-JavaDoc'

myathk committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/TaskList.java

myathk committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/TaskList.java

myathk committed:
Merge branch 'add-gradle-support' of https://github.com/NUSmhk/ip

myathk committed:
Add Gradle support

myathk committed:
Merge tag 'A-Gradle'

myathk committed:
Add CheckStyle and resolve CheckStyle errors in code

myathk committed:
Merge branch 'branch-A-Checkstyle'

myathk committed:
Add Basic GUI

myathk committed:
Rename command package to commands and refactor

myathk committed:
Rewrite code to return String and show in GUI instead

myathk committed:
Merge branch 'branch-Level-10'

kflim commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,74 @@
+import duke.TaskList;
+import duke.ToDo;
+import duke.Ui;
+
```
Maybe you could add some header comments to the class and the public methods? I noticed most of the Command files are missing these comments as well, perhaps because they are newly committed files

myathk replied:
You are right, thank you for pointing it out! I have only added header comments for half of my class and public methods and I am still working towards completing for all of them.

kflim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+    private static String detectedFormat;
+    private static LocalTime startTime;
+    private static LocalTime endTime;
+    private static final String[] dateTimeFormats = {
```
I like the attention to detail, nice job!

kflim reviewed (commented):
In general, the code is compliant with the coding standard. Just some tweaks for some files and it will be good!

kflim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        "yyyy/MM/dd HHmm",
+        "yyyy-MM-dd HHmm",
+        "dd MMM yyyy hh:mm a"
+    };
```
I like the attention to detail, nice job!

myathk replied:
Thank you!

myathk committed:
Add header comments

myathk committed:
Delete Ui class and refactor code

myathk committed:
Add header comments for command package classes

jayasting98 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate date;
+    protected LocalDateTime dateTime;
```
Since if I am not wrong, there are no external package subclasses that require these two variables, perhaps you can consider making these private? On the other hand, even if there were any such subclasses, maybe you could consider making these private regardless and use accessor methods instead? I noticed this in the `Event` class as well.

myathk replied:
Thank you! I just found out the difference between private and protected so I will change them accordingly.

jayasting98 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        "hh:mm a"
+    };
+
+    /**
+     * This method takes in the user input as a String and runs checks on it to determine
+     * the type of command it is, also throwing exception if the input is invalid.
+     * It also takes in a TaskList that Duke bot uses to load and
+     * store tasks the user adds or deletes.
+     * @param fullCommand String of user input.
+     * @param taskList a list to store tasks the user adds or deletes.
+     * @return Command which Duke bot can execute afterwards.
+     * @throws DukeException if user input cannot be understood.
+     */
+    public static Command parse(String fullCommand, TaskList taskList) throws DukeException {
```
While I do like that your Javadoc header comments are very descriptive, I think it may be desirable to have a short first sentence as a summary (usually one line). Maybe you could consider opting to use more subsequent sentences for describing in more detail. I noticed this in other places as well, though it is a minor issue and it is definitely better that you are descriptive (which I really like) than having no description at all.

myathk replied:
I also just found out that a shorter description seems to be the norm so I will stick to that in the future, thank you!

jayasting98 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,43 @@
+     * @param startTime start time of Event.
+     * @param endTime end time of Event.
+     */
+    public Event(String input, LocalDate date, LocalTime startTime, LocalTime endTime) {
```
Is there not supposed to be an empty line between the description and the parameter section in the Javadoc header comment? I noticed this in your other Javadoc header comments as well, though it is a minor issue.

myathk replied:
You are correct, I seem to have missed out on that, it is indeed a violation of the coding standard. Thanks for pointing it out, I will change them.

jayasting98 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,43 @@
+        return String.format("[E]%s (at: %s from: %s)", super.toString(),
+                this.date.format(DateTimeFormatter.ofPattern("dd MMM yyyy")),
+                    this.startTime.format(DateTimeFormatter.ofPattern("hh:mm a")) + "-"
+                        + this.endTime.format(DateTimeFormatter.ofPattern("hh:mm a")));
```
Maybe you can consider [using intermediate variables](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-complicated-expressions) to make this clearer.

myathk replied:
Thank you, I've changed it.

jayasting98 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+                throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+            }
+        }
+    }
```
Perhaps you could consider extracting several sub-methods from this to [avoid this pretty long method](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods).

myathk replied:
Thank you! You are right, it is my mistake. I have improved it.

jayasting98 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+        } else {
+            throw new DukeException("OOPS!!! The duration of an event cannot be empty.");
+        }
+    }
```
Is this [amount of nesting](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting) necessary? Maybe you can consider moving each nested `if` block out of its parent `if` and then put it after the parent `if` block since you can take advantage of the thrown exceptions to make subsequent statements unreachable as intended. Not sure if that makes sense so here's an example.
From this:
```java
if (isCondition1) {
    // everything within these braces is unreachable if isCondition1 is false
    if (isCondition2) {
        // ...
    } else {
        // ...
    }
} else {
    // ...
    throw new DukeException("...");
}
```
To this:
```java
if (isCondition1) {
    // ...
} else {
    // ...
    throw new DukeException("...");
}
// everything after this is still unreachable if isCondition1 is false because
// the exception would then be thrown before this
if (isCondition2) {
    // ...
} else {
    // ...
}
```

myathk replied:
Thank you for your clear explanation, it helped me understand the necessary changes which I've implemented!

jayasting98 reviewed (commented):
Overall, pretty good code, just a few minor issues. I really love the amount of detail you go to such as checking for many possible cases for exceptions. I also like how descriptive your Javadoc header comments are.

myathk committed:
Change access modifiers of class fields to private

myathk committed:
Fix header comments coding standard violation

myathk committed:
Fix deep nesting of error handling code

myathk committed:
Refactor parse method into several sub-methods to shorten it

myathk committed:
Add intermediate variables to toString method

myathk committed:
Add assertions to files where needed

Duke does not have assertions to test assumptions and implementing
one will be good to allow indication of possible bugs.

Add assertions in
* private methods to check parameter validity
* default switch cases
* for-loop when getting task from tasklist

In this situation, assertion can help identify any illegal parameters in
private methods, ensure no default cases in a switch statement occurs
and have for-loops run only when necessary.

myathk committed:
Refactor to improve code quality where necessary

Parser class has poor code quality with complicated expressions, long methods
and numerous exception messages.

Refactoring of code is necessary to improve readability of code so that further
implementations can be done with more ease.

Some complicated expressions in Parser has been refactored to have
more constant values and steps. SLAP has been used to refactor a section of parse function
into a new parseFind function to handle FindCommand. All exception messages are
refactored to having different types recognised by a String instead of using full messages.

These abstractions will help improve readability of code.

myathk committed:
Merge pull request #1 from NUSmhk/branch-A-Assertions

Add assertions to files where needed

myathk committed:
Merge branch 'master' of https://github.com/NUSmhk/ip

myathk committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/TaskList.java

myathk committed:
Set up Continuous Integration

myathk committed:
Refactor AddCommand class into 3 new classes

myathk committed:
Add support for Deadline type task to have recurring tasks

myathk committed:
Merge branch 'branch-B-RecurringTasks'

myathk committed:
Set Duke's personality to be a grumpy Terminator.

myathk committed:
Improve GUI

myathk committed:
Add Command command to get a list of commands that are available to user

myathk committed:
Add exception handling for empty task list and add a representative screenshot

myathk committed:
Add clear Command to delete all tasks

myathk committed:
Add clear Command into list of Commands available

myathk committed:
Change file name

myathk committed:
Add User Guide

myathk committed:
Set theme jekyll-theme-hacker

myathk committed:
Fix typo

myathk committed:
Fix typo

myathk committed:
Merge branch 'master' of https://github.com/NUSmhk/ip

myathk committed:
Add credit for reused code

myathk committed:
Add quick start guide

myathk committed:
Update main page README.md
